static/js/place/chart_embed.tsx

```diff
@@ -0,0 +1,176 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from "react";
+import {
+  Button,
+  Modal,
+  ModalHeader,
+  ModalBody,
+  ModalFooter,
+} from "reactstrap";
+import { randDomId, saveToFile } from "../shared/util";
+
+interface ChartEmbedStateType {
+  modal: boolean;
+  chartDom: Node;
+  svgHtml: string;
+  dataCsv: string;
+}
+
+/**
+ * A component to include with each Chart, that displays embed information and data for a chart
+ * in a Modal
+ */
+class ChartEmbed extends React.Component<unknown, ChartEmbedStateType> {
+  private textareaElement: React.RefObject<HTMLTextAreaElement>;
+  private modalId: string;
+
+  constructor(props: unknown) {
+    super(props);
+    this.state = {
+      modal: false,
+      svgHtml: "",
+      chartDom: null,
+      dataCsv: "",
+    };
+    this.modalId = randDomId();
+    this.textareaElement = React.createRef();
+    this.toggle = this.toggle.bind(this);
+    this.onOpened = this.onOpened.bind(this);
+    this.copySvgToClipboard = this.copySvgToClipboard.bind(this);
+    this.downloadData = this.downloadData.bind(this);
+    this.onClickTextarea = this.onClickTextarea.bind(this);
+  }
+
+  /**
+   * Toggles the view state of the modal.
+   */
+  public toggle(): void {
+    this.setState({
+      modal: !this.state.modal,
+    });
+  }
+
+  /**
+   * Updates the view state of the modal to true, and includes the data necessary for displaying the modal.
+   */
+  public show(svgHTML: string, chartDOM: Node, dataCsv: string): void {
+    this.setState({
+      modal: true,
+      svgHtml: svgHTML,
+      chartDom: chartDOM,
+      dataCsv: dataCsv,
+    });
+  }
+
+  /**
+   * Callback for after the modal has been rendered and added to the DOM.
+   */
+  public onOpened(): void {
+    const modalElement = document.getElementById(this.modalId);
```

Comment:
use react ref for this element?

Reply:
that didn't work, mostly because i couldn't resolve the type for RefObject<Modal<T>> with the right T to get the rest of the code working. so i updated it to do things the old school way